### PR TITLE
BLD: Define PyErr_SetFromWindowsErr on Cygwin.

### LINKED ISSUE #22997: [Bug]: Cygwin build fails due to use of Windows-only functions in _tkagg.cpp
### Bug summary

When the build gets to https://github.com/matplotlib/matplotlib/blob/main/src/_tkagg.cpp#L262-L273 on Cygwin, the build fails with a few `goto crosses initialization` warnings, which are easy to fix (closed by #23051), and two `error: ‘PyErr_SetFromWindowsErr’ was not declared in this scope`, which are less easy to fix.

### Code for reproduction

```python
pip install matplotlib
```

### Actual outcome

```con
...
lots of compiler command lines and some warnings (`array subscript has type 'char'`)
...
          gcc ...  -I/usr/include/python3.8 -c src/_tkagg.cpp -o build/temp.cygwin-3.3.4-x86_64-3.8/matplotlib.backends._tkagg/src/_tkagg.o -fvisibility=hidden -flto
          src/_tkagg.cpp: In function ‘void load_tkinter_funcs()’:
          src/_tkagg.cpp:263:9: error: ‘PyErr_SetFromWindowsErr’ was not declared in this scope
            263 |         PyErr_SetFromWindowsErr(0);
                |         ^~~~~~~~~~~~~~~~~~~~~~~
          src/_tkagg.cpp:271:9: error: ‘PyErr_SetFromWindowsErr’ was not declared in this scope
            271 |         PyErr_SetFromWindowsErr(0);
                |         ^~~~~~~~~~~~~~~~~~~~~~~
          src/_tkagg.cpp:283:1: error: jump to label ‘exit’
            283 | exit:
                | ^~~~
          src/_tkagg.cpp:272:14: note:   from here
            272 |         goto exit;
                |              ^~~~
          src/_tkagg.cpp:274:26: note:   crosses initialization of ‘bool tk_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |                          ^~~~~
          src/_tkagg.cpp:274:10: note:   crosses initialization of ‘bool tcl_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |          ^~~~~~
          src/_tkagg.cpp:283:1: error: jump to label ‘exit’
            283 | exit:
                | ^~~~
          src/_tkagg.cpp:268:14: note:   from here
            268 |         goto exit;
                |              ^~~~
          src/_tkagg.cpp:274:26: note:   crosses initialization of ‘bool tk_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |                          ^~~~~
          src/_tkagg.cpp:274:10: note:   crosses initialization of ‘bool tcl_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |          ^~~~~~
          src/_tkagg.cpp:283:1: error: jump to label ‘exit’
            283 | exit:
                | ^~~~
          src/_tkagg.cpp:264:14: note:   from here
            264 |         goto exit;
                |              ^~~~
          src/_tkagg.cpp:274:26: note:   crosses initialization of ‘bool tk_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |                          ^~~~~
          src/_tkagg.cpp:274:10: note:   crosses initialization of ‘bool tcl_ok’
            274 |     bool tcl_ok = false, tk_ok = false;
                |          ^~~~~~
          error: command '/usr/bin/gcc' failed with exit code 1
...
pip reminds me this is not pip's fault
 ```

### Expected outcome

```con
...
Successfully installed collected packages: matplotlib
```

### Additional information

Cygwin 3.3.4
GCC 11.2.0
Python 3.8.12

I've installed matplotlib in earlier versions.  #17415 rather suggests that Cygwin builds worked at that time, and probably made using the default backend somewhat easier.  #22445 seems to be where the `PyErr_SetFromWindowsErr` calls came in.

I've attempted to create a dummy `PyErr_SetFromWindowsErr` on Cygwin, but, given the number of failures in the test suite, I probably didn't do it right.

### Operating system

Cygwin 3.3.4 on Windows 10

### Matplotlib Version

3.5.2

### Matplotlib Backend

TkAgg

### Python version

3.8.12

### Jupyter version

N/A

### Installation

pip

### PR DESCRIPTION
**Pull request by DWesl**
## PR Summary

Define `PyErr_SetFromWindowsErr` on Cygwin.  This function is defined on Windows, not Cygwin.

Closes #22997.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [X] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [X] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: src/_tkagg.cpp [hunk 1]
```diff
@@ -24,6 +24,9 @@
  * also.
  */
 #define WIN32_DLL
+static inline PyObject *PyErr_SetFromWindowsErr(int ierr) {
+    return PyErr_SetString(PyExc_OSError, "Call to EnumProcessModules failed");
```
Comment (anntzer) [inline src/_tkagg.cpp:28]: You can't `return PyErr_SetString(...)` as the latter returns void.  So it's probably rather `PyErr_SetString(...); return NULL;`.

**Comment by anntzer** [comment]
@DWesl just as curiosity, if you can build cpython from source yourself, can you check whether changing the ifdef at https://github.com/python/cpython/blob/39a54ba63850e081a4a5551a773df5b4d5b1d3cd/Include/pyerrors.h#L188 and https://github.com/python/cpython/blob/a4460f2eb8b9db46a9bce3c450c8b038038a7c93/Python/errors.c#L867 to also define these functions on cygwin works?

Also, should the blocks protected by `ifdef _WIN32` in _c_internal_utils.c also likewise be also enabled on cygwin (with similar fixes for the uses of PyErr_SetFromWindowsErr)?

**Comment by anntzer** [comment]
No hurries.

**Comment by DWesl** [comment]
It looks like it compiles, but it's a bit tricky to get useful test results with the system FreeType, rather than specifically FreeType 2.6.1

**Comment by jklymak** [comment]
Is this still a work-in-progress?  Moved to draft, but feel free to ask that it be moved back for active review...

**Review by QuLogic — APPROVED** [review]
This successfully fixes build on Cygwin; not sure there is anything else that really needs to be done.

**Review by anntzer — APPROVED** [review]

**Pull request merged**